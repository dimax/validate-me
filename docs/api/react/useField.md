---
sidebarDepth: -1
---

# useField

## Props

| Name           |               Type               | Required |
| -------------- | :------------------------------: | :------: |
| props          |              Object              |   Yes    |
| props.form     |               Form               |   Yes    |
| props.name     |              string              |   Yes    |
| props.rules    |         `Array<string>`          |    No    |
| props.value    |               any                |    No    |
| props.required |             boolean              |    No    |
| props.type     | `'number' | 'checkbox' | 'date'` |    No    |

## Returns

An array of two objects

### First element: field state

| Name                |  Type   |
| ------------------- | :-----: |
| fieldState          | Object  |
| fieldState.pristine | boolean |
| fieldState.loading  | boolean |
| fieldState.error    | string  |
| fieldState.warning  | string  |

### Second element: input props

| Name                |   Type   |
| ------------------- | :------: |
| inputProps          |  Object  |
| inputProps.required | boolean  |
| inputProps.name     |  string  |
| inputProps.type     |  string  |
| inputProps.value    |   any    |
| inputProps.onChange | Function |
| inputProps.onBlur?  | Function |
