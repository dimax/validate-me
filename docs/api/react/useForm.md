---
sidebarDepth: -1
---

# useForm

## Named exports

### `VContext`

Sets the form instance into the validateme context to allow the fields to access it. It's mandatory to use it.

### `setErrorHandler(handler: (error: any) => Array<RawRules>): void`

Configures the error handler for parsing the errors thrown by the server.

## Props

Does not receives any props.

## Returns

The form instance.

#### Form instance public API

- `validate(): [boolean, Object]`: If the form is pristine, it touches every field. If the validation success, returns `[true, Object]`, where the `Object` is the representation of the inputs and its values. If the validation fails, it returns `[false]`.
- `process(error: any): void`: Processes the server error and injects the new errors and warnings to every failed field.

#### Form instance "private" API

The same object has the following props used _only_ by `useField`.

- `setField(name: string, field: Field): void`
