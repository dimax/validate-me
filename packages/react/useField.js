import { useEffect, useReducer, useMemo, useRef } from 'react';
import getRules from '@validate-me/core/getRules';
import { loadRule, processRawRules } from '@validate-me/core/rules';
import { getMessage, getWarning } from '@validate-me/core/dictionary';

const touchedAction = ['touched', true];
const notLoadingAction = ['loading'];
const loadingAction = ['loading', true];
const warning = ['warning'];

export default function useField(
  type,
  { form, rules, value, name, required, min, max, pattern, multiple, options },
) {
  const checkbox = type === 'checkbox';
  const prop = checkbox ? 'checked' : 'value';
  const checkList = checkbox && options;
  const stateRef = useRef();
  const [ruleRunners, setRules] = useReducer(
    (rules, rule) => (rule.length == null ? rule : rules.concat(rule)),
    [],
  );
  const [state, setState] = useReducer(
    (state, prop) => {
      const oldVal = state.value;
      const [key, val] = typeof prop === 'function' ? prop(oldVal) : prop;

      if (state[key] === val) {
        return state;
      }

      const newState = { ...state, [key]: val };

      if (
        key === 'value' &&
        ruleRunners.length &&
        oldVal !== val &&
        (val || required)
      ) {
        newState.error = '';

        for (const rule of ruleRunners) {
          if (!rule.run(val)) {
            newState.error = getMessage(rule, val);

            break;
          }
        }
      }

      return newState;
    },
    {
      loading: true,
      value: value || (checkList ? {} : ''),
    },
  );
  const onChange = useMemo(() => {
    if (checkList) {
      return evt => {
        const { defaultValue, checked } = evt.target;

        setState(value => ['value', { ...value, [defaultValue]: checked }]);
      };
    }

    return evt => {
      setState(['value', evt.target[prop]]);
    };
  }, [checkList, prop]);

  useEffect(
    () =>
      form.setField(name, {
        touch() {
          setState(touchedAction);

          return stateRef.current.error;
        },
        invalid() {
          return Boolean(stateRef.current.loading || stateRef.current.error);
        },
        clearWarning() {
          setState(warning);

          return stateRef.current.value;
        },
        parseError: rawError =>
          loadRule(rawError)
            .then(rule => {
              setRules(rule);
              setState(value => ['error', getMessage(rule, value)]);
            })
            .catch(rule => {
              if (rule instanceof Error) {
                throw rule;
              }
              setState(value => ['warning', getWarning(rule, value)]);
            }),
      }),
    [name, form],
  );

  useEffect(() => {
    setState(loadingAction);

    const baseRules = getRules(type, {
      required,
      min,
      max,
      pattern,
      multiple,
    });

    processRawRules(rules ? baseRules.concat(rules) : baseRules, setRules, () =>
      setState(notLoadingAction),
    );
  }, [max, min, multiple, pattern, required, rules, type]);

  useEffect(() => {
    stateRef.current = state;
  }, [state]);

  return [
    state,
    {
      required,
      name,
      type,
      min,
      max,
      pattern,
      multiple,
      onChange,
      [prop]: state.value,
      // onBlur.once
      onBlur: state.touched ? undefined : () => setState(touchedAction),
    },
  ];
}
