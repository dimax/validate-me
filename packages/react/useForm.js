import { useReducer, useEffect, useCallback, useRef } from 'react';
import { processErrors } from '@validate-me/core/rules';

let errorHandler = f => f;
const INIT_STATE = { invalid: true };
const invalidAction = ['invalid', true];
const validAction = ['invalid'];
const touchAction = ['touched', true];

export function handleState(state, [type, value]) {
  return state[type] === value
    ? state
    : {
        ...state,
        [type]: value,
      };
}

export function setErrorHandler(handler) {
  errorHandler = handler;
}

export default function useForm() {
  const fields = useRef({});
  const [state, setState] = useReducer(handleState, INIT_STATE);
  const [fieldStates, setFieldState] = useReducer(handleState, {});
  const setField = useCallback((name, field) => {
    fields.current[name] = field;
  }, []);
  // Executes the submit event. Touches every pristine input.
  const validate = useCallback(() => {
    let localInvalid;
    const fieldsValue = Object.values(fields.current);

    if (!state.touched) {
      localInvalid = fieldsValue.reduce(
        (invalid, field) => field.touch() || invalid,
        true,
      );

      setState(touchAction);
    }

    if (localInvalid || state.invalid) {
      return false;
    }

    fieldsValue.forEach(field => field.clearWarning());

    return true;
  }, [state.invalid, state.touched]);
  // Process errors from server
  const process = useCallback(
    error => processErrors(fields.current, errorHandler(error)),
    [],
  );

  // Verifies if every input is valid
  useEffect(() => {
    for (const fieldInvalid of Object.values(fieldStates)) {
      if (fieldInvalid) {
        setState(invalidAction);

        return;
      }
    }

    setState(validAction);
  }, [fieldStates]);

  return [state, { setField, setFieldState, validate, process }];
}
