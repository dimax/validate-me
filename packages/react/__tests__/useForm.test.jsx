import React from 'react';
import { renderHook, cleanup, act } from 'react-hooks-testing-library';

import useField from '../useField';
import useForm, { setErrorHandler, VContext } from '../useForm';

function formRenderer(form) {
  // eslint-disable-next-line react/prop-types
  const FormContext = ({ children }) => (
    <VContext.Provider value={form}>{children}</VContext.Provider>
  );

  return FormContext;
}

describe('react/useForm', () => {
  const inputName = 'text';
  const props = {
    name: inputName,
    pattern: '^\\D+$',
    rules: [['len', '1', '5']],
  };
  const renderInput = FormContext =>
    renderHook(() => useField('text', props), { wrapper: FormContext });

  afterEach(() => {
    cleanup();
  });

  test('Instanciates a form and validates a valid and invalid input states', async () => {
    const { result: form } = renderHook(() => useForm());
    const FormContext = formRenderer(form.current);
    const { result: field, waitForNextUpdate: fieldWait } = renderInput(
      FormContext,
    );

    await act(fieldWait);

    act(() => {
      field.current[1].onChange({ target: { value: 'a' } });
    });
    expect(form.current.validate()).toMatchSnapshot();

    act(() => {
      field.current[1].onChange({ target: { value: '1' } });
    });
    expect(form.current.validate()).toMatchSnapshot();
  });

  test('Instanciates a form and process errors from server with(out) custom error handler', async () => {
    const { result: form } = renderHook(() => useForm());
    const FormContext = formRenderer(form.current);
    const { result: field, waitForNextUpdate: fieldWait } = renderInput(
      FormContext,
    );

    await act(fieldWait);

    await act(() => form.current.process({ [inputName]: ['required'] }));
    expect(field.current[0]).toMatchSnapshot();
    expect(form.current.validate()).toMatchSnapshot();

    await act(async () => {
      setErrorHandler(err => err.fields);
      await form.current.process({
        code: 1,
        fields: { [inputName]: ['warningFromServer'] },
      });
      setErrorHandler(f => f);
    });
    act(() => {
      field.current[1].onChange({ target: { value: 'a' } });
    });
    expect(field.current[0]).toMatchSnapshot();
    expect(form.current.validate()).toMatchSnapshot();
  });
});
