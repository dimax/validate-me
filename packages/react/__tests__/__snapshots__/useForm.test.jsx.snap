// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react/useForm Instanciates a form and process errors from server with(out) custom error handler 1`] = `
Object {
  "error": "The field is required.",
  "loading": undefined,
  "value": "",
}
`;

exports[`react/useForm Instanciates a form and process errors from server with(out) custom error handler 2`] = `
Array [
  false,
]
`;

exports[`react/useForm Instanciates a form and process errors from server with(out) custom error handler 3`] = `
Object {
  "error": "",
  "loading": undefined,
  "touched": true,
  "value": "a",
  "warning": "WARNING: Unknown server rule \\"warningFromServer\\" failed with the value \\"\\"",
}
`;

exports[`react/useForm Instanciates a form and process errors from server with(out) custom error handler 4`] = `
Array [
  true,
  Object {
    "text": "a",
  },
]
`;

exports[`react/useForm Instanciates a form and validates a valid and invalid input states 1`] = `
Array [
  true,
  Object {
    "text": "a",
  },
]
`;

exports[`react/useForm Instanciates a form and validates a valid and invalid input states 2`] = `
Array [
  false,
]
`;
